### ISSUE TITLE
Task creation/modification in admin is a huge security hole

### ISSUE THREAD

**Issue by olivierdalang** [Koed00/django-q#371]
Hi !

I think it's very dangerous to allow to freely create/modify tasks from the admin. This allows to very easily execute arbitrary code for anyone who can gain access to that part of the admin and is also very error-prone.

The ability to monitor and enable/disable tasks from the admin is really nice, but those tasks would definitely have to be predefined in the code, not modifiable through the UI.

**Comment by quantumlink** [comment]
Not sure I understand. The Admin is protected by both permissions and authentication. You don't have to give full access to the admin. I don't see this as a security hole and certainly can't see it as a "huge security hole"

**Comment by olivierdalang** [comment]
It allows execution of arbitrary python, which is quite an escalation from getting access to the admin account.

The admin isn't necessarily well secured (session left open by users, weak password, permissions not well defined, etc...).

**Comment by maerteijn** [comment]
Checking for `DEBUG` mode isn't a well thought of solution in my opinion as you would disable very useful functionality in a production environment.

You can already configure all of this as you wish with the build in permission system Django provides, so it is not allowed to add / edit those specific model types, in this case tasks for specific or all users ).

That the admin is not "well secured" has also nothing to do with Django-Q: You can secure the admin as tight as you want, including 2FA or firewall rules or whatever you feel you want to secure the admin. When unwanted admin access occurs, the creation/modification of tasks is the least of your problems then.

**Comment by carltongibson** [comment]
It’s also worth saying that you’re free to unregister any models you don’t want to appear in the admin.

**Comment by olivierdalang** [comment]
The admin not being always secure is of course out of scope of django-q, but the escalation from admin account to running arbitrary code is not. I don't think there's any other place in django where this happens ?

Checking for DEBUG or unregister the models altogether is indeed not ideal. I think a proper fix would be to register tasks in the code, and only allow to run those tasks rather than any callable.

**Comment by markopy** [comment]
I agree with @olivierdalang that his serious security issue and arbitrary task creation should not be possible by default. This should be behind a specific setting with a big fat warning. There are a lot of ways credentials can be lost or admin permissions misconfigured. In no way should this lead to trivial server take over and blanket access to the entire database.
If the default django admin had this problem it would be considered a grade A vulnerability.

The easiest way for users of this project to secure themselves until this is fixed is to unregister the schedule admin:
```
from django_q.admin import Schedule
admin.site.unregister(Schedule)
```

**Comment by stormlifter** [comment]
I see the ideal solution is the to register functions that will show up as options on the scheduled task list drop down in the admin.

**Comment by markopy** [comment]
@stormlifter Yes, that would be a secure implementation of this feature provide parameters to these functions a appropriately sanitized. This is still a big if and requires a clear warning.

**Comment by stormlifter** [comment]
I don't agree with a warning, developers write software and ultimately it is them who secure it, no need for a warning if a developer is the one who registered the function for use. I even think the sanitizing is the responsibility of the developer too, but I'd accept just string argument values as a compromise especially since it simplifies the implementation.

If we're talking about having unlimited time to develop, then the functions could have Django patterned forms and each argument handled by an appropriate form field.

**Comment by markopy** [comment]
I meant a warning in the documentation for developers who want to register functions for scheduling. As evidenced by this thread a lot of people don't understand the implications at all.

**Comment by stormlifter** [comment]
I dunno, I think folks get it, just don't give a User access to Scheduling that you don't want to also give access to running arbitrary code, within my teams that's just developers that have access, so the same people that could just update the code to make it do anything already. It's a lot of power for sure, probably too much, but its not a security hole without a solution that's accepted within Django to secure the whole site.

Your claim that it's a show-stopping security hole is a bit dramatic all things considered. Simply do not give admin users access to Scheduler if you have any concerns. 

However we've outlined a clear way forward if we want improvements, so who's gonna write the code and make the PR? :) 

I eventually might but right now within my use it hasn't become high priority as the same people who write the code have access to that in the Admin. To allow end-users and non-developers to configure the scheduled tasks we use an API with serialization and appropriate handling there.

**Comment by olivierdalang** [comment]
Well I made my point already, but would definitely suggest to make the admin opt-in with a big red warning in the docs until this is addressed.

Looking at this repo's stat, you can be certain not all of the users of this package are as careful as you and your users seem to be...

**Comment by carltongibson** [comment]
It’s worth noting too that making a user Staff doesn’t thereby grant them permissions to any particular admin. You need to grant them further permissions. Sure, if you make them a superuser, but then they’re a superuser... The “hole” here is already opt-in for this reason. 

@stormlifter’s take seems the correct one. If you’re careful with permissions, which you will be if you’re at all concerned, this isn’t a major issue. (Maybe a warning in the docs, but months later not seeing a PR.)

**Comment by olivierdalang** [comment]
@carltongibson just for my general knowledge, is there any other place in Django or major extension where it's also possible to escalate to execute arbitrary code in a similar way ?

**Comment by carltongibson** [comment]
Did Django-Q exist before it was written? 

This is a feature. Don’t give it to just anybody sure. But that’s true of the entire admin. It’s not something extra special to panic about.

**Comment by markopy** [comment]
@carltongibson I think the point where we differ here is that in my opinion a misconfiguration or failure in django's permission system should never lead to a server compromise by allowing arbitrary code execution. This is very much a defense in depth issue. As a django user I expect django permissions to protect different domains inside my web application, anything that allows escalation beyond that is a vulnerability.

Now I understand the point that django users are developers and there is a lot of freedom to shoot yourself in the foot. That's totally fine. What is not appropriate IMO is that this is enabled by default. If someone enables this with a setting after reading the attached warning in the docs this whole ticket would not have to exist.

You have to accept that this is a very widely used project based on the github stats and I can guarantee most people who install it never considered what is discussed here in any way. It is the responsibility of the project maintainers to provide secure defaults for these people.

As for a PR, I should point out that I'm not even a user of DjangoQ at this point. I found this issue during evaluation and posted a helpful mitigation for people who don't need the schedule admin at all. This issue has been open for almost a year and the response @olivierdalang received is astonishing to me.

I also think his PR tying the fix to `DEBUG` is not great, but making the schedule admin read only unless changed by a setting is the easiest right thing to do. If you provide secure defaults with a warning the responsibility falls to the individual developer. As it stands, it's DjangoQ's.

**Comment by carltongibson** [comment]
Yeah... I guess I don’t see “arbitrary code execution” as any more scary than “delete all your users” or “access all their data”, which permissions errors in the admin already allow. 

If we can’t assume you set permissions in the admin correctly then I don’t think we can say that it’s safe to use it at all.

**Comment by markopy** [comment]
That's a very weird threat model to have. It sounds like you would also consider SQL injection not worth fixing as long as it only appears limited to admin users?
Any django install that's more than a small project will be part of a larger system where code execution leads to further compromise.

**Comment by carltongibson** [comment]
An SQL injection that was only performance via the admin would be lower priority that one that was accessible from say a public view yes. But that’s not analogous. Here we’re allowing trusted users to create arbitrary tasks. You may not want that, which is fine, don’t use the admin here, but it’s not some scary security hole. Labelling it “arbitrary code execution” doesn’t make it so.

Presumably you’d only grant this access to your dev team, and they can already do what they like anyway. 

I can see that in some environments you’d want more separation, but then again you wouldn’t use the admin here. For the vast majority of Django apps this is a feature, a good one. 

Clearly we don’t agree...

**Comment by markopy** [comment]
Clearly we don't, but I just want to make sure we have the same understanding of what "arbitrary" means here. It's an arbitrary code execution because there is no limitations on what kind of code can be executed, not because it can be executed by arbitrary users. [Arbitrary code execution](https://en.wikipedia.org/wiki/Arbitrary_code_execution) is a term of art.

**Comment by olivierdalang** [comment]
@carltongibson @stormlifter is it actually clear that you can open django-q's admin, and in `func` and `args` enter whatever python callable you want ? Including things like **DONT TRY THAT** func=`os.system` args=`"rm -rf ."` ?

Obviously one can get more creative than just deleting everything and inject password loggers, ransomware, spambots, etc...

How is that not some scary security hole ?

**Comment by carltongibson** [comment]
Yes. Thanks. I understand. 😀

My point is more that so-labelling it doesn’t change the situation. It’s no more an issue because you give it a bad name. 

Rather, still, again, it’s a feature, and it’s already “opt-in”. You’re taking about making someone a superuser, or granting them access explicitly here. Your comments seem to suggest it’s the former that’s the issue: but then you made them a superuser, if you didn’t mean that you already have more troubles than you can count.

**Comment by olivierdalang** [comment]
1. Granting some of my users access to django-q's admin, to be able to monitor task execution any maybe rerun some tasks. Just like I also gave them access to user management, or modification logs... Sounds sensible.

2. Granting those same users ssh access to the server would never cross my mind. Even if I know they are well intentionned.

Now, while doing 1/, it appears I'm actually also doing 2/. How many of the 257 declared users of this app are aware of this ?

I don't get how you can be cool with that ?!

I'm going to stop after this one, happy that at least I'm not completely alone being surprised by this :-)

**Comment by stormlifter** [comment]
It sounds to me like you gotta build a non-admin interface for your users, maybe only give them read access for Scheduler? Both of those seem reasonable. Perhaps see giving access (especially write permissions) to the admin as truly admin access, just make an interface for users where you do all the security specifically to the use.

**Comment by markopy** [comment]
@stormlifter That's all fine but the issue is that if you gave someone admin access in django they have admin access. After installing this extension with default settings they now also effectively have ssh access to your server. This is not documented anywhere and very much unexpected.

Anyway, @olivierdalang and myself have explained the issue and I'm satisfied you understand the implications. Do with that what you will.

**Comment by carltongibson** [comment]
Not granting superuser access to all staff is the starting point. As a rule of thumb try just giving that to those you would grant ssh access to. Then give out permission as are needed, as is the intention of the admin's use of the permissions system. 

This issue has been open nearly a year, and no PR.
